spec/controllers/saml_idp_controller_spec.rb

```diff
@@ -1675,217 +1629,142 @@ def name_id_version(format_urn)
       end
     end
 
-    context 'nameid_format is missing' do
+    describe 'NameID format' do
       let(:user) { create(:user, :fully_registered) }
+      let(:subject_element) { xmldoc.subject_nodeset[0] }
+      let(:name_id) { subject_element.at('//ds:NameID', ds: Saml::XML::Namespaces::ASSERTION) }
+      let(:auth_settings) { saml_settings(overrides: { name_identifier_format: }) }
+      let(:email_allowed) { nil }
+      let(:name_identifier_format) { nil }
 
       before do
         stub_analytics
-        allow(@analytics).to receive(:track_event)
-      end
-
-      it 'defaults to persistent' do
-        auth_settings = saml_settings(overrides: { name_identifier_format: nil })
         service_provider = build(:service_provider, issuer: auth_settings.issuer)
         IdentityLinker.new(user, service_provider).link_identity
-        user.identities.last.update!(verified_attributes: ['email'])
-        generate_saml_response(user, auth_settings)
+        ServiceProvider.
```

Comment:
i would pull the service_provider out into a `let`, then you can do updates in describe blocks without having to do a query

```
let(:service_provider) do
  build(:service_provider, issuer: auth_settings.issuer, email_nameid_format_allowed: email_allowed)
end
```

Reply:
That makes sense. Unfortunately, when I tried it, I found issues with issuer uniqueness, certificate validation and seeded test data. Trying to clean this up...